File: scala/space-age/src/main/scala/SpaceAge.scala
object SpaceAge {
  val agesOnPlanets: Map[String, Double] = Map(
    "Mercury" -> 0.2408467,
    "Venus" -> 0.61519726,
    "Earth" -> 1.0,
    "Mars" -> 1.8808158,
    "Jupiter" -> 11.862615,
    "Saturn" -> 29.447498,
    "Uranus" -> 84.016846,
    "Neptune" -> 164.79132,
  )
  val earthYear = 31_557_600

  def onEarth(age: Double): Double =
    age / earthYear

  def onVenus(age: Double): Double =
    onEarth(age) / agesOnPlanets("Venus")

  def onMercury(age: Double): Double =
    onEarth(age) / agesOnPlanets("Mercury")

  def onMars(age: Double): Double =
    onEarth(age) / agesOnPlanets("Mars")

  def onJupiter(age: Double): Double =
    onEarth(age) / agesOnPlanets("Jupiter")

  def onSaturn(age: Double): Double =
    onEarth(age) / agesOnPlanets("Saturn")

  def onUranus(age: Double): Double =
    onEarth(age) / agesOnPlanets("Uranus")

  def onNeptune(age: Double): Double =
    onEarth(age) / agesOnPlanets("Neptune")
}

File: kotlin/space-age/src/main/kotlin/SpaceAge.kt
class SpaceAge(
    val seconds: Int,
) {
    private enum class Planet(
        val orbitalPeriod: Double,
    ) {
        EARTH(1.0),
        MERCURY(0.2408467),
        VENUS(0.61519726),
        MARS(1.8808158),
        JUPITER(11.862615),
        SATURN(29.447498),
        URANUS(84.016846),
        NEPTUNE(164.79132),
    }

    private val earthYearSeconds = 31_557_600

    private val earthYears = seconds.toDouble() / earthYearSeconds

    fun onEarth(): Double = earthYears

    fun onMercury(): Double = earthYears / Planet.MERCURY.orbitalPeriod

    fun onVenus(): Double = earthYears / Planet.VENUS.orbitalPeriod

    fun onMars(): Double = earthYears / Planet.MARS.orbitalPeriod

    fun onJupiter(): Double = earthYears / Planet.JUPITER.orbitalPeriod

    fun onSaturn(): Double = earthYears / Planet.SATURN.orbitalPeriod

    fun onUranus(): Double = earthYears / Planet.URANUS.orbitalPeriod

    fun onNeptune(): Double = earthYears / Planet.NEPTUNE.orbitalPeriod
}
